Feature/asarray

oleksandr-pavlyk opened the pull request:
Closes #640 . Adds `dpctl.tensor.asarray` and `dpctl.tensor.empty`.

```python
dpctl.tensor.empty(shape, dtype="f8", order="C", device=None, usm_type="device", sycl_queue=None)
```

Create an empty `dpctl.tensor.usm_ndarray` object of requested `dtype` and based on request USM allocation type on specified array-API `device`, and of requested layout specified by `order` keyword.

Instead of specifying array-API device keyword, `sycl_queue` can be used to specify the target of allocation instead.

```python
dpctl.tensor.asarray(obj, dtype=None, copy=None, order="K", device=None, usm_type=None, sycl_queue=None)
```

Converts Python object `obj` to `usm_ndarray` object. Support objects are:

1. `usm_ndarray` instances
2. `numpy.ndarray` instances
3. Python objects exposing `__sycl_usm_array_interface__`.
4. Python sequences of Python scalars

reazulhoque commented:
@oleksandr-pavlyk Do we have a test where we are testing the equality of shape of new array created using `asarray` and the shape is multi-dimensional? Can you please point me to it?

oleksandr-pavlyk commented:
Is this good enough?

```
In [1]: import dpctl, dpctl.tensor as dpt

In [2]: dpt.asarray([[1, 2, 3], [4,5,6]], dtype="d").shape
Out[2]: (2, 3)

In [3]: X = dpt.asarray([[1, 2, 3], [4,5,6]], dtype="d", device="cpu", usm_type="device")

In [4]: Y = dpt.asarray(X, device="gpu", usm_type="shared")

In [5]: X.shape == Y.shape
Out[5]: True

In [6]: X.sycl_device
Out[6]: <dpctl.SyclDevice [backend_type.opencl, device_type.cpu,  Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz] at 0x7f8836ab93b0>

In [7]: Y.sycl_device
Out[7]: <dpctl.SyclDevice [backend_type.level_zero, device_type.gpu,  Intel(R) UHD Graphics [0x9bca]] at 0x7f8858a49870>
```

PokhodenkoSA commented:
The description of this PR has more documentation than function docstrings you created in code.

densmirn commented on dpctl/tensor/_ctors.py:
```diff
@@ -0,0 +1,377 @@
+    """Represents object `obj` as usm_ndarray"""
+    # 1. Check that copy is a valid keyword
+    if copy not in [None, True, False]:
+        raise TypeError(
```
Seems `ValueError` is more suitable here.

oleksandr-pavlyk replied:
I disagree, it my mind the check is asking if `copy` is of `bool` type, or of `None` type.

densmirn commented on dpctl/tensor/_ctors.py:
```diff
@@ -0,0 +1,377 @@
+):
+    """Creates empty usm_ndarray"""
+    dtype = np.dtype(dtype)
+    if len(order) == 0 or order[0] not in "CcFf":
```
`order=None` will raise `TypeError`.

densmirn commented on dpctl/tensor/_ctors.py:
```diff
@@ -0,0 +1,377 @@
+    copy=None,
+    usm_type=None,
+    sycl_queue=None,
+    order="K",
```
Why is there `order="K"` by default? `numpy` has `order="C"` by default.

densmirn replied:
Sorry numpy has `order=None` by default.

oleksandr-pavlyk replied:
`numpy.asarray` has [default value](https://numpy.org/doc/stable/reference/generated/numpy.asarray.html) of `order` set to `"C"`. The `numpy.array` [uses different default](https://numpy.org/doc/stable/reference/generated/numpy.array.html) of `order="K"`.

Array API's `asarray` [does not provision](https://data-apis.org/array-api/latest/API_specification/creation_functions.html#asarray-obj-dtype-none-device-none-copy-none) `order` keyword.

I can change the default to `"C"`.

densmirn commented on dpctl/tensor/_ctors.py:
```diff
@@ -0,0 +1,377 @@
+
+
+def empty(
+    sh, dtype="f8", order="C", device=None, usm_type="device", sycl_queue=None
```
Parameter `usm_type="device"` by default for `empty`, but the parameter is `None` by default for `asarray`. Is it designed so?

oleksandr-pavlyk replied:
Yes. For `asarray( another_usm_array, usm_type=None)` will infer the `usm_type` from the input. If the input is not an object with `__sycl_usm_array_interface__`, the `usm_type` will be assume to be `"device"`.

densmirn commented on dpctl/tensor/_ctors.py:
```diff
@@ -0,0 +1,377 @@
+    obj,
+    dtype=None,
+    device=None,
+    copy=None,
```
`copy=None` looks like `copy=True` in the code, but `numpy.asarray` tries to not create copy. I`m not sure, but maybe we need to change the parameter default value to False?

oleksandr-pavlyk replied:
`copy=None` should be more like `copy=False`, rather than `copy=True`.

oleksandr-pavlyk commented:
@diptorupd @densmirn @reazulhoque @Alexander-Makaryev @PokhodenkoSA 

This PR is now feature complete and changed from draft to "ready for review". 
I added docstrings, added CHANGELOG, added tests (guides by coverage).

I cleaned up commit history to reduce it from 32 commits to 10 commit,
I do **not** consider it a good use of our time to split this PR into any number of smaller PRs.

oleksandr-pavlyk commented:
> The description of this PR has more documentation than functions docstrings you created in code.

I added docstrings.

diptorupd commented on dpctl/tensor/_copy_utils.py:
```diff
@@ -93,12 +93,12 @@ def copy_to_numpy(ary):
 
 
-def copy_from_numpy(np_ary, usm_type="device", queue=None):
+def copy_from_numpy(np_ary, usm_type="device", sycl_queue=None):
```
`sycl_queue`? Are you sure we will not be changing it again in couple of months 😄

oleksandr-pavlyk replied:
Maybe, but it has to stay consistent across API surface (we have `device` and `sycl_queue` in `asarray`, but `queue` in `from_numpy`).

reazulhoque reviewed (approved):
Tested manually with DuckUSMArray to make sure integration with Numba-dppy is working correctly.
